otherlibs/systhreads/Makefile

```diff
@@ -13,96 +13,107 @@
-libthreadsnat.a: $(NATIVECODE_C_OBJS)
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
+allopt: lib$(LIBNAME)nat.$(A) $(LIBNAME).cmxa $(CMIFILES)
 
-st_stubs_n.o: st_stubs.c st_posix.h
-	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
-	            $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
-	            -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
-	mv st_stubs.o st_stubs_n.o
+lib$(LIBNAME).$(A): $(BYTECODE_C_OBJS)
+	$(MKLIB) -o $(LIBNAME) $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
 
-threads.cma: $(THREAD_OBJS)
-	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
-	  -cclib -lunix $(PTHREAD_CAML_LINK)
+lib$(LIBNAME)nat.$(A): $(NATIVECODE_C_OBJS)
+	$(MKLIB) -o $(LIBNAME)nat $^
+
+$(LIBNAME).cma: $(THREADS_BCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix -linkall $(PTHREAD_CAML_LINK) $^
+# TODO: Figure out why -cclib -lunix is used here.
+# It may be because of the threadsUnix module which is deprecated.
+# It may hence be good to figure out whether this module shouldn't be
+# removed, and then -cclib -lunix arguments.
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $(PTHREAD_CAML_LINK) $^
+endif
 
 # See remark above: force static linking of libthreadsnat.a
-threads.cmxa: $(THREAD_OBJS:.cmo=.cmx)
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+	$(CAMLOPT) -linkall -a -cclib -lthreadsnat $(PTHREAD_CAML_LINK) -o $@ $^
 
 # Note: I removed "-cclib -lunix" from the line above.
 # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+st_stubs_b.$(O): st_stubs.c $(HEADER)
+	$(BYTECC) -I$(ROOTDIR)/byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) -c $<
+	mv st_stubs.$(O) $@
 
-$(THREAD_OBJS:.cmo=.cmx): ../../ocamlopt
+st_stubs_n.$(O): st_stubs.c $(HEADER)
+	$(NATIVECC) -I$(ROOTDIR)/asmrun -I$(ROOTDIR)/byterun $(NATIVECCCOMPOPTS) \
+	  $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
+	  -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c $<
+	mv st_stubs.$(O) $@
 
 partialclean:
 	rm -f *.cm*
 
 clean: partialclean
-	rm -f *.o *.a *.so
+	rm -f *.$(A) *.$(O)
 
 INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
+THREADS_LIBDIR=$(INSTALL_LIBDIR)/$(LIBNAME)
 INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
 
 install:
-	if test -f dllthreads.so; then \
-	  cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
-	cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
-	cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
-	if test -d $(INSTALL_LIBDIR)/threads; then :; \
-	  else mkdir $(INSTALL_LIBDIR)/threads; fi
-	cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
-	rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
-	cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
-	   $(INSTALL_LIBDIR)
-	cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
+	cp libthreads.$(A) $(INSTALL_LIBDIR)
```

Comment:
Alain Frisch (2016/09/09 06:18 -0700):
> AppVeyor fails with
> 
> ````
> cp libthreads.lib C:/Program Files/OCaml/lib
> cp: target 'Files/OCaml/lib' is not a directory
> ````
 
Alain Frisch (2016/09/09 06:19 -0700):
> >  INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
> >  
> >  install:
> > -	if test -f dllthreads.so; then \
> > -	  cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
> > -	cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
> > -	cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
> > -	if test -d $(INSTALL_LIBDIR)/threads; then :; \
> > -	  else mkdir $(INSTALL_LIBDIR)/threads; fi
> > -	cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
> > -	rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
> > -	cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
> > -	   $(INSTALL_LIBDIR)
> > -	cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
> > +	cp libthreads.$(A) $(INSTALL_LIBDIR)
> 
> Missing quotes around the target directory, I guess.
 
Alain Frisch (2016/09/09 06:20 -0700):
> >  
> >  installopt:
> > -	cp libthreadsnat.a $(INSTALL_LIBDIR)/libthreadsnat.a
> > -	cd $(INSTALL_LIBDIR); $(RANLIB) libthreadsnat.a
> > -	cp $(THREAD_OBJS:.cmo=.cmx) threads.cmxa threads.a \
> > -	   $(INSTALL_LIBDIR)/threads
> > -	cd $(INSTALL_LIBDIR)/threads && $(RANLIB) threads.a
> > +	cp libthreadsnat.$(A) "$(INSTALL_LIBDIR)"
> > +	cd $(INSTALL_LIBDIR); $(RANLIB) libthreadsnat.$(A)
> > +	cp $(THREADS_NCOBJS) threads.cmxa threads.$(A) $(THREADS_LIBDIR)
> 
> Missing quotes as well (also on `cd` lines above and below).

Indeed, sorry about that and thanks a lot for having reported it.

This should now be fixed and the PR has been rebased on the latest
trunk.